mitmproxy/websocket.py

```diff
@@ -1,168 +1,136 @@
 """
-*Deprecation Notice:* Mitmproxy's WebSocket API is going to change soon,
-see <https://github.com/mitmproxy/mitmproxy/issues/4425>.
+Mitmproxy used to have its own WebSocketFlow type until mitmproxy 6, but now WebSocket connections now are represented
+as HTTP flows as well. They can be distinguished from regular HTTP requests by having the
+`mitmproxy.http.HTTPFlow.websocket` attribute set.
+
+This module only defines the classes for individual `WebSocketMessage`s and the `WebSocketData` container.
 """
-import queue
 import time
 import warnings
 from typing import List
 from typing import Optional
-from typing import Union
 
-from mitmproxy import flow
+from mitmproxy import stateobject
 from mitmproxy.coretypes import serializable
-from mitmproxy.net import websocket
-from mitmproxy.utils import human
-from mitmproxy.utils import strutils
-
-from wsproto.frame_protocol import CloseReason
-from wsproto.frame_protocol import Opcode
 
 
 class WebSocketMessage(serializable.Serializable):
     """
-    A WebSocket message sent from one endpoint to the other.
+    A single WebSocket message sent from one peer to the other.
+
+    Fragmented WebSocket messages are reassembled by mitmproxy and the
+    represented as a single instance of this class.
+
+    The [WebSocket RFC](https://tools.ietf.org/html/rfc6455) specifies both
+    text and binary messages. To avoid a whole class of nasty type confusion bugs,
+    mitmproxy stores all message contents as binary. If you need text, you can decode the `content` property:
+
+    >>> if message.is_text:
+    >>>     text = message.content.decode()
     """
 
-    type: Opcode
-    """indicates either TEXT or BINARY (from wsproto.frame_protocol.Opcode)."""
     from_client: bool
     """True if this messages was sent by the client."""
-    content: Union[bytes, str]
+    is_text: bool
+    """
+    True if the message is a text message, False if the message is a binary message.
+
+    In either case, mitmproxy will store the message contents as *bytes*.
+    """
+    content: bytes
```

Comment:
What's the reason we're switching to bytes for text messages as well? We know that WebSocket text is UTF-8, doesn't that make it more awkward to work with? Or is that my lack of Python?

Reply:
I haven't really made my mind up here to be honest. The previous implementation was annoying me quite a bit because `content` was `Union[str,bytes]`, and you always had to add a special case and check. Now my hope is that it's just more consistent to deal with, and if you need text you can just call `.decode()` yourself. @Kriechi, any opinions from your end?

Reply:
Déjà vu

https://github.com/mitmproxy/mitmproxy/issues/4131

> I remotely recall something about WebSocket TEXT frames having a very strictly defined encoding? It looks like we just forward what we get from wsproto. 

---

> and you always had to add a special case and check

How is that special case gone? You still need to decide when to `.decode()` and when not to.

I'm not favoring one solution or the other, but given that I've brought this exact topic up before maybe we can finally settle this :wink:

Reply:
I swear this is the last comment: maybe my original idea of a `text` property kind of makes sense? It would just be a getter/setter that does the decode/encode for you. Otherwise every user would require the knowledge that WebSocket text frames are always utf-8 to manually do the `decode` of the bytes.

If `not is_text` the `text` property should crash or something tho? Because reading `text` for arbitrary `content` won't always work.